### PR TITLE
Optimize MoE routing for Gemma4

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Add an optional direct-token-gather path for sparse MoE routing. It gathers input tokens directly in expert order instead of materializing Top-K copies with jnp.repeat before sorting.

For Gemma4-26B, this removes the large Top-8 activation expansion while preserving expert-ID rotation and correct gradient accumulation.

The optimization is controlled by moe_use_direct_token_gather, which defaults to false and is currently enabled only for Gemma4-26B. Its performance improvement has only been validated on that model; other MoE configurations retain the existing repeat-and-sort path until they are benchmarked.

# Tests

Add unit test for numerical equivalence.

XProf confirms that the direct-gather path removes the broadcast_bitcast_fusion associated with the repeated activation, and improved end-to-end step time.
Before: [xprof](https://xprof.corp.google.com/trace_viewer/aireenmei-13534665124925283610?view_start=11352.617&view_end=11358.416), [screenshot](https://screenshot-v2.corp.google.com/40jao38emg6eo), step time: 10.5 s
Before: [xprof](https://xprof.corp.google.com/trace_viewer/aireenmei-6309662714480977089?view_start=11644.359&view_end=11652.992), [screenshot](https://screenshot-v2.corp.google.com/718p0f3ruv7uo), step time: 10.39 s

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -952,11 +961,8 @@ def permute(
       if roll_to_expert_id is not None:
         flatten_selected_experts = (flatten_selected_experts - roll_to_expert_id) % self.num_experts
       sorted_selected_experts = jnp.argsort(flatten_selected_experts)
-      # sort inputs for number of selected experts
-      replicated_inputs_2d = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
-      sorted_inputs = _sort_activations(replicated_inputs_2d, sorted_selected_experts, use_custom_sort_vjp).astype(
-          self.dtype
-      )
+      routing_experts = flatten_selected_experts.reshape(selected_experts.shape)
+      sorted_inputs = _route_activations(inputs_2d, routing_experts).astype(self.dtype)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/layers/moe.py]: update as suggested

File: tests/unit/moe_test.py [hunk 2]
```diff
@@ -504,6 +504,34 @@ def fake_shard_map(function, *, mesh, in_specs, out_specs, check_vma):
...
+
+    if compute_gradient:
+      cotangent = jnp.arange(48, dtype=jnp.float32).reshape(12, 4)
+      expected = jax.grad(lambda x: jnp.sum(reference(x) * cotangent))(inputs)
+      result = jax.grad(lambda x: jnp.sum(actual(x) * cotangent))(inputs)
+    else:
+      expected = reference(inputs)
+      result = actual(inputs)
+
+    np.testing.assert_array_equal(result, expected)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/unit/moe_test.py]: This test uses small integer-valued float32 arrays, and only performs indexing and sums of exactly representable integers. There should be no floating-point approximation difference. Exact equality is better here.

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -952,11 +963,13 @@ def permute(
       if roll_to_expert_id is not None:
         flatten_selected_experts = (flatten_selected_experts - roll_to_expert_id) % self.num_experts
       sorted_selected_experts = jnp.argsort(flatten_selected_experts)
-      # sort inputs for number of selected experts
-      replicated_inputs_2d = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
-      sorted_inputs = _sort_activations(replicated_inputs_2d, sorted_selected_experts, use_custom_sort_vjp).astype(
-          self.dtype
-      )
+      if self.config.moe_use_direct_token_gather:
+        sorted_inputs = _route_activations(inputs_2d, flatten_selected_experts).astype(self.dtype)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:967]: if sorted_selected_experts is not used here, maybe move line 954 `sorted_selected_experts = jnp.argsort(flatten_selected_experts)` in the else branch?

File: src/maxtext/layers/moe.py [hunk 4]
```diff
@@ -151,6 +152,16 @@ def _sort_activations(
...
 
 
+def _route_activations(inputs: jax.Array, selected_experts: jax.Array) -> jax.Array:
+  """Groups token activations by expert without materializing Top-K copies."""
+  selected_experts = selected_experts.reshape((inputs.shape[0], -1))
+  # When use_gather_mosaic_kernel=False, use the general JAX backward, which
+  # gathers and sums gradients from all expert copies of each token. The
+  # optimized Mosaic gather-reduce kernel currently requires exactly 8 selected
+  # experts per token and is enabled only by the specialized batch-split path.
+  return sort_activations.route(inputs, selected_experts, use_gather_mosaic_kernel=False)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:162]: if use_gather_mosaic_kernel is fixed as false when moe_use_direct_token_gather=True, we should add a check in types.py to throw early value error

**Review by NuojCheng — APPROVED** [review]
LGTM to me with some minor comments. Note that this change does not affect the ragged sort path.

File: src/maxtext/configs/models/gemma4-26b.yml [hunk 5]
```diff
@@ -46,6 +46,7 @@ num_experts_per_tok: 8
 shared_experts: 1
 norm_topk_prob: true
 load_balance_loss_weight: 0.001
+moe_use_direct_token_gather: true
```
Comment (RissyRan) [inline src/maxtext/configs/models/gemma4-26b.yml]: I think this is an implementation feature (perf config), but model specific config (like num_experts etc)? Recommend to remove it if this is the case.

File: tests/unit/moe_test.py [hunk 6]
```diff
@@ -504,6 +504,34 @@ def fake_shard_map(function, *, mesh, in_specs, out_specs, check_vma):
 class RoutedMoeTest(parameterized.TestCase):
   """Routed Mixture of Experts test."""
 
+  @parameterized.parameters(False, True)
+  def test_route_activations_matches_repeat_and_sort(self, compute_gradient):
```
Comment (RissyRan) [inline tests/unit/moe_test.py]: I think it's better to directly test `_route_activations()` with moe_use_direct_token_gather=true/false?

File: tests/unit/moe_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/unit/moe_test.py]: +1 to `It is safer and more robust to use np.testing.assert_allclose with appropriate tolerances.` Also, let's use float32 for inputs instead of integer? This should be more similar to real situation.

**Review by RissyRan — COMMENTED** [review]
Thanks! Minor comments

File: src/maxtext/layers/moe.py (hunk 3 above)
Comment (aireenmei) [reply src/maxtext/layers/moe.py:967]: Although not used in the `self.config.moe_use_direct_token_gather` branch. sorted_selected_experts is later returned by this function on line 1008, so it has to be outside the else branch.

File: src/maxtext/layers/moe.py (hunk 4 above)
Comment (aireenmei) [reply src/maxtext/layers/moe.py:162]: added early validation

File: src/maxtext/configs/models/gemma4-26b.yml (hunk 5 above)
Comment (aireenmei) [reply src/maxtext/configs/models/gemma4-26b.yml]: Makes sense, removed

File: tests/unit/moe_test.py (hunk 6 above)
Comment (aireenmei) [reply tests/unit/moe_test.py]: I think you mean testing moe_use_direct_token_gather=true/false at the `permute()` function level since `_route_activations()` is a helper function for the moe_use_direct_token_gather=true path. I've updated the test for this purpose and renamed to `test_permute_direct_token_gather_matches_repeat_and_sort`

**Review by RissyRan — APPROVED** [review]
